### PR TITLE
[inductor] Misc division lowering fixes

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88938
* #88936
* __->__ #88603
* #88937

1. `aten.div.Tensor_mode` should allow broadcasting
2. `div` can use `ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT`
3. `prims.div` on integers should be truncating division
4. Add lowering for `true_divide` which is aliased to `div`
5. register lowering for inplace version of `div_mode`

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by ngimel** [comment]
I'd prefer to not touch division unless we have a good reason to. Reference division is overcomplicated and not necessarily performance friendly, changing all division ops just to get corner cases of fp division with `floor` rounding seems like too much.

**Comment by peterbell10** [comment]
> changing all division ops just to get corner cases of fp division with floor rounding seems like too much.

Is `rounding_mode=None` the main concern here, or is floating point floor division a real concern?

**Comment by ngimel** [comment]
The concern here is swapping correct and performant integer floordiv https://github.com/pytorch/pytorch/pull/88603/files#diff-ea3834aeda2e22f49dd5ba0becd456b449b10d17dfe64c34e89072701c82873dL281 for performance unknown change from the references, and yes, performance unknown changes to `rounding_mode=None`.

**Comment by peterbell10** [comment]
Here is a before and after comparison of the triton code generated for `dtype, rounding_mode in product((int64, float32), (None, "floor"))`

## `int64`, `rounding_mode=None`

Before
```python
    tmp0 = tl.load(in_ptr0 + (x0), xmask)
    tmp2 = tl.load(in_ptr1 + (x0), xmask)
    tmp1 = tmp0.to(tl.float32)
    tmp3 = tmp2.to(tl.float32)
    tmp4 = tmp1 / tmp3
    tl.store(out_ptr0 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp4, xmask)
```

After
```python
    tmp0 = tl.load(in_ptr0 + (x0), xmask)
    tmp2 = tl.load(in_ptr1 + (x0), xmask)
    tmp1 = tmp0.to(tl.float32)
    tmp3 = tmp2.to(tl.float32)
    tmp4 = tmp1 / tmp3
    tl.store(out_ptr0 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp4, xmask)
```

This is completely identical

## `float32`, `rounding_mode=None`

Before
```
    tmp0 = tl.load(in_ptr0 + (x0), xmask)
    tmp1 = tl.load(in_ptr1 + (x0), xmask)
    tmp2 = tmp0 / tmp1
    tl.store(out_ptr0 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp2, xmask)
```

After
```
    tmp0 = tl.load(in_ptr0 + (x0), xmask)
    tmp1 = tl.load(in_ptr1 + (x0), xmask)
    tmp2 = tmp0 / tmp1
    tl.store(out_ptr0 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp2, xmask)
```

Again, identical.

## `int64`, `rounding_mode="floor"`

Before 
```
    tmp0 = tl.load(in_ptr0 + (x0), xmask)
    tmp1 = tl.load(in_ptr1 + (x0), xmask)
    tmp2 = tl.where((tmp0 < 0) != (tmp1 < 0), tl.where(tmp0 % tmp1 != 0, tmp0 // tmp1 - 1, tmp0 // tmp1), tmp0 // tmp1)
    tl.store(out_ptr0 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp2, xmask)
```

After
```
    tmp0 = tl.load(in_ptr0 + (x0), xmask)
    tmp1 = tl.load(in_ptr1 + (x0), xmask)
    tmp2 = tmp0 // tmp1
    tmp3 = tl.libdevice.signbit(tmp0) if (tmp0).dtype is tl.float32 else tmp0 < 0
    tmp4 = tl.libdevice.signbit(tmp1) if (tmp1).dtype is tl.float32 else tmp1 < 0
    tmp5 = tmp3 != tmp4
    tmp6 = tmp0 % tmp1
    tmp7 = 0
    tmp8 = tmp6 != tmp7
    tmp9 = tmp5 & tmp8
    tmp10 = tmp9.to(tl.int64)
    tmp11 = tmp2 - tmp10
    tl.store(out_ptr0 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp11, xmask)
```

While this one is spread out over more lines, it is almost identical. The only differences are:
1. `signbit` expands  to an if-statement. However, this should be optimized away as the condition is on the type and for integers it evaluates to `(x < 0)` which is what `ops.floordiv` uses.
2. instead of `where(cond1, where(cond2, x - 1, x), x)`, the decomposition calculates `x - (cond1 and cond2).to(int64)`

In terms of runtime performance, I get both as taking exactly `85.3 us` for 1 million elements so I would say they are functionally identical.

## `float32`, `rounding_mode="floor"`

This is where the real change in this PR is, and of course it doesn't look great but that's because python's floating point floor division function is rather cumbersome.

Before
```
    tmp0 = tl.load(in_ptr0 + (x0), xmask)
    tmp1 = tl.load(in_ptr1 + (x0), xmask)
    tmp2 = tmp0 / tmp1
    tmp3 = tl.libdevice.floor(tmp2)
    tl.store(out_ptr0 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp3, xmask)
```

After
```
    tmp0 = tl.load(in_ptr0 + (x0), xmask)
    tmp1 = tl.load(in_ptr1 + (x0), xmask)
    tmp2 = tl.libdevice.fmod(tmp0, tmp1)
    tmp3 = 0.0
    tmp4 = tmp2 != tmp3
    tmp5 = tmp0 < tmp3
    tmp6 = tmp1 < tmp3
    tmp7 = tmp5 ^ tmp6
    tmp8 = tmp4 & tmp7
    tmp9 = tmp0 - tmp2
    tmp10 = tmp9 / tmp1
    tmp11 = 1.0
    tmp12 = tmp10 - tmp11
    tmp13 = tl.where(tmp8, tmp12, tmp10)
    tmp14 = tmp13 != tmp3
    tmp15 = tl.libdevice.floor(tmp13)
    tmp16 = tmp13 - tmp15
    tmp17 = 0.5
    tmp18 = tmp16 > tmp17
    tmp19 = tmp15 + tmp11
    tmp20 = tl.where(tmp18, tmp19, tmp15)
    tmp21 = tmp0 / tmp1
    tmp22 = tl.libdevice.signbit(tmp21) if (tmp21).dtype is tl.float32 else tmp21 < 0
    tmp23 = 0
    tmp24 = tl.abs(tmp23)
    tmp25 = -tmp24
    tmp26 = tl.where(tmp22, tmp25, tmp24)
    tmp27 = tl.where(tmp14, tmp20, tmp26)
    tmp28 = tmp1 != tmp3
    tmp29 = tl.where(tmp28, tmp27, tmp21)
    tl.store(out_ptr1 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp29, xmask)
```

In terms of performance, I see a minor slowdown from 42 us to 43.4 µs for 1 million elements. Though the real downside will be in limiting the number of pointwise operations that can be fused with this kernel. However, this is necessary if we want to give the same results as eager mode and, by extension, numpy and python's floor division.

**Comment by ngimel** [comment]
The real downside is indeed number of the pointwise ops that will be fused in the kernel, and we already had serious perf regressions when that was tweaked #87189. If the only existing problem is floor division for floats, I'd rather tolerate small differences with eager.

**Comment by ngimel** [comment]
Additionally, since references go through primtorch broadcasting, it hardcodes broadcast sizes and interferes with guards that inductor usually generates for pointwise ops (with symints it's probably becoming better, but I'm not sure of the current status, cc @chillee).

**Comment by peterbell10** [comment]
> The real downside is indeed number of the pointwise ops that will be fused in the kernel, and we already had serious perf regressions when that was tweaked https://github.com/pytorch/pytorch/pull/87189. If the only existing problem is floor division for floats, I'd rather tolerate small differences with eager.

Okay.

> Additionally, since references go through primtorch broadcasting, it hardcodes broadcast sizes and interferes with guards that inductor usually generates for pointwise ops

So as a general rule, any decomposition not specifically written for inductor will have this problem. Is there any way we can fix symint support without completely rewriting them for inductor?

**Comment by Chillee** [comment]
> it hardcodes broadcast sizes and interferes with guards that inductor usually generates for pointwise ops

I'm not sure what you mean by guards in this case? Generally, since we specialize on 1, we're not gonna have any guards anyways.

Like, I think for static shapes, this shouldn't cause any issues right?

**Comment by ngimel** [comment]
For static - yeah, it doesn't matter, for dynamic - inductor generates equality for corresponding shapes (and thus replaces them with the same symbol). 
But generally, inductor handles broadcasting itself, and it's a reasonable thing for it to do, whereas refs insist on doing broadcasting, and there's no way to turn it off. 
This is an issue mostly for primitive elementwise ops that inductor already has lowerings for, so I don't see it as preventing use of references in general (and inductor uses a lot of references from _refs).

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -3364,7 +3364,7 @@ def truncdiv(a, b):
     return ops.truncdiv(a, b)
 
 
-@register_lowering(aten.div.Tensor_mode)
+@register_lowering(aten.div, broadcast=True)
```
Comment (ngimel) [inline torch/_inductor/lowering.py:3367]: can you please add a test case for this (and other) fixes?

**Comment by peterbell10** [comment]
@ngimel PTAL. I've added tests that fail on master for me.

**Review by ngimel — APPROVED** [review]
Thanks!

**Pull request closed without merging**